File: src/AdventOfCode2024.Common/FileService.fs
namespace AdventOfCode2024.Common

open System
open System.IO
open System.Text.RegularExpressions

module FileService =

    let getFileAsString (fileName: string) : string =
        File.ReadAllText($"Data/{fileName}")

    let getFileAsArray (fileName: string) (delimiterPattern: string) : seq<string> =
        File.ReadAllText($"Data/{fileName}")
        |> fun contents -> Regex.Split(contents, delimiterPattern)
        |> Seq.filter (fun line -> not (String.IsNullOrWhiteSpace(line)))

    let getFileAs2dIntArray (fileName: string) (rowDelimiterPattern: string) : int[,] =
        let rows = 
            File.ReadAllText($"Data/{fileName}")
            |> fun contents -> Regex.Split(contents, rowDelimiterPattern)
            |> Array.filter (fun row -> not (String.IsNullOrWhiteSpace(row)))

        if rows.Length = 0 then
            Array2D.zeroCreate 0 0
        else
            let height = rows.Length
            let width = rows.[0].Length
            Array2D.init height width (fun i j -> int (rows.[i].[j].ToString()))

    let getFileAsListOfListOfInt (fileName: string) (rowDelimiterPattern: string) : list<list<int>> =
        let rows = 
            File.ReadAllText($"Data/{fileName}")
            |> fun contents -> Regex.Split(contents, rowDelimiterPattern)
            |> Array.filter (fun row -> not (String.IsNullOrWhiteSpace(row)))

        if rows.Length = 0 then
            []
        else
            rows
            |> Array.map (fun row -> row.Split(" ") |> Array.map int |> Array.toList)
            |> Array.toList

File: src/AdventOfCode2024.Common.CSharp/FileService.cs
using System.Text.RegularExpressions;

namespace AdventOfCode2024.Common.CSharp;

public static class FileService
{
    public static string GetFileAsString(string fileName)
    {
        var contents = File.ReadAllText($"Data/{fileName}");

        return contents;
    }

    public static IEnumerable<string> GetFileAsArray(string fileName, string delimiterPattern = @"\r\n|\n")
    {
        var contents = File.ReadAllText($"Data/{fileName}");
        return Regex.Split(contents, delimiterPattern).Where(line => !string.IsNullOrEmpty(line));
    }

    public static int[,] GetFileAs2dIntArray(string fileName, string rowDelimiterPattern = @"\r\n|\n")
    {
        var rows = Regex.Split(File.ReadAllText($"Data/{fileName}"), rowDelimiterPattern).Where(row => !string.IsNullOrEmpty(row)).ToArray();

        if (rows.Length == 0)
        {
            return new int[0, 0];
        }

        var response = new int[rows.Length, rows[0].Length];

        for (var i = 0; i < rows.Length; i++)
        {
            var row = rows[i].ToCharArray();

            for (var j = 0; j < row.Length; j++)
            {
                response[i, j] = int.Parse(row[j].ToString());
            }
        }

        return response;
    }

    public static List<List<int>> GetFileAsListOfListOfInt(string fileName, string rowDelimiterPattern = @"\r\n|\n")
    {
        var rows = Regex.Split(File.ReadAllText($"Data/{fileName}"), rowDelimiterPattern).Where(row => !string.IsNullOrEmpty(row)).ToArray();

        return rows.Length == 0 ?
            new() : rows.Select(t => t.Split(" ")).Select(row => row.Select(int.Parse).ToList()).ToList();
    }

    public static char[,] GetFileAs2dCharArray(string fileName)
    {
        var lines = GetFileAsArray(fileName).ToArray();
        int rows = lines.Length;
        int cols = lines[0].Length;

        var grid = new char[rows, cols];
        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                grid[i, j] = lines[i][j];
            }
        }
        return grid;
    }
}
